apps/server/src/modules/tldraw/service/tldraw.ws.service.ts

```diff
@@ -136,84 +155,174 @@ export class TldrawWsService {
 					}
 					break;
 				case WSMessageType.AWARENESS: {
-					applyAwarenessUpdate(doc.awareness, decoding.readVarUint8Array(decoder), conn);
+					const update = decoding.readVarUint8Array(decoder);
+					this.publishUpdateToRedis(doc, update, 'awareness');
 					break;
 				}
 				default:
 					break;
 			}
 		} catch (err) {
-			doc.emit('error', [err]);
+			this.logger.warning(new WebsocketMessageErrorLoggable(err));
+			throw err;
 		}
 	}
 
+	/**
+	 * @param {{ Buffer }} channel redis channel
+	 * @param {{ Buffer }} update update message
+	 * @param {{ WsSharedDocDo }} doc ydoc
+	 */
+	public redisMessageHandler = (channel: Buffer, update: Buffer, doc: WsSharedDocDo): void => {
+		const channelId = channel.toString();
+
+		if (channelId === doc.name) {
+			applyUpdate(doc, update, this.sub);
+		}
+
+		if (channelId === doc.awarenessChannel) {
+			applyAwarenessUpdate(doc.awareness, update, this.sub);
+		}
+	};
+
 	/**
 	 * @param {WebSocket} ws
 	 * @param {string} docName
 	 */
-	public setupWSConnection(ws: WebSocket, docName = 'GLOBAL'): void {
+	public async setupWSConnection(ws: WebSocket, docName: string) {
 		ws.binaryType = 'arraybuffer';
+
 		// get doc, initialize if it does not exist yet
-		const doc = this.getYDoc(docName, true);
-		doc.conns.set(ws, new Set());
+		const isNew = !this.docs.has(docName);
+		const doc = this.getYDoc(docName);
+		doc.connections.set(ws, new Set());
+		if (isNew) {
+			await this.updateDocument(docName, doc);
+		}
+
+		ws.on('error', (err) => {
+			this.logger.warning(new WebsocketErrorLoggable(err));
+		});
 
-		// listen and reply to events
 		ws.on('message', (message: ArrayBufferLike) => {
 			this.messageHandler(ws, doc, new Uint8Array(message));
 		});
 
-		// Check if connection is still alive
+		// check if connection is still alive
 		let pongReceived = true;
 		const pingInterval = setInterval(() => {
-			const hasConn = doc.conns.has(ws);
-
-			if (pongReceived) {
-				if (!hasConn) return;
+			if (pongReceived && doc.connections.has(ws)) {
 				pongReceived = false;
-
-				try {
-					ws.ping();
-				} catch (e) {
-					this.closeConn(doc, ws);
-					clearInterval(pingInterval);
-				}
+				ws.ping();
 				return;
 			}
 
-			if (hasConn) {
-				this.closeConn(doc, ws);
-			}
-
+			void this.closeConn(doc, ws);
```

Comment:
await instead of void by all  void this.closeConn(doc, ws); or?

Reply:
we are not able to use await in this context.

Reply:
we do not want to await any Redis specific actions, thats why they are marked as void, they were writtern as hanging promises before but I guess you wanted us to change them to async await?

Reply:
Ok, but then you must add a error handling for the unhandled promise chain. Otherwise a error by closing the connection run into unhandledPromiseRejections Error what is one of the worst cases in nodejs. It is also marked from sonar default settings as high code smell.